### ISSUE TITLE
Design rationale

### ISSUE THREAD

**Issue by patrickbkr** [MARTIMM/gnome-source-skim-tool#1]
Hi! This is more of a question than an issue. (If the following is offensive to you in any way, then please tell. It's not meant to offend. I'm genuinely interested in your insights.)

Some time ago I have read a bit into the [GObject Introspection](https://gi.readthedocs.io/en/latest/#gobject-introspection) project and how they promote creating bindings for it. I also put some thought into how this could be done in Raku. From what I've read they promote using the [libgirepository](https://gi.readthedocs.io/en/latest/writingbindings/libgirepository.html) C library to introspect the library at hand *at runtime* and dynamically generate the required classes, modules and functions during `use`.

There is a [list](https://gi.readthedocs.io/en/latest/users.html) of bindings, many of those follow this runtime approach (not all). I've looked at the Perl variant, [Glib::Object::Introspection](https://gitlab.gnome.org/GNOME/perl-glib-object-introspection). It's not that much code after all. (I admit mostly C, but I believe most if not all could be plain Raku.)

From what I've heard from Xliff, compile times are a big issue with static binding libraries of the size of GTK. A runtime binding wouldn't have that issue.

Now there are quite some Raku GTK binding projects: GTK::Simple, Xliffs bindings, yours. None of which went with the GObject Introspection approach.

Can you give some insight into your rationale? Is GObject Introspection an bad idea for some reason? Why is no one going with this approach?

**Comment by MARTIMM** [comment]
Well, I didn't, sorry. The way it goes always; The mail is going into a map and I check it regularly until there is no mail for ages and then forget to check it and all of a sudden there is mail lying there waiting to be opened, sorry :-|

I have made a small program to see what it can do. It looks nice but there are some problems I find difficult to explain. A small example of a problem is accepting a series of arguments from the caller program and then comparing it with the results from the call to the GIR libs before attempting to call the routine the user wants to be called.

Below is this small program I made to see what can be done with the GIR libs. It does not do much. Do some tests, show some info, and list the functions in the gtk_label class.

I also did not find the routines to get documentation out which is needed to inform the user of the libraries. Therefore I decided to use the XML files (in `/usr/share/gir-1.0/`) and generate Raku code from it including the docs which is still an ongoing process.

However, I want to try to use those GIR libraries too, when I feel up to it, just like you said, it should be faster. The docs can still be generated the way I do it now including renaming the C-variables etc.

```
use NativeCall;

use Gnome::N::N-GObject;
use Gnome::N::GlibToRakuTypes;

use Gnome::Glib::Error;

#-------------------------------------------------------------------------------
constant \GIREPOSITORY = 'libgirepository-1.0.so';
constant \Error = Gnome::Glib::Error;


#-------------------------------------------------------------------------------
my N-GObject $repository = g_irepository_get_default;

my $e = CArray[N-GError].new(N-GError);
my N-GObject $typelib = g_irepository_require(
  $repository, 'Gtk', '3.0', 0, $e
);
if $typelib {
  note 'typelib ok';
}

else {
  my Error() $error = $e[0];
  note $error.message;
}

# libs are ok
note 'lib: ', g_irepository_get_shared_library( $repository, 'Gtk');

# works! version is 3.0
note 'version: ', g_irepository_get_version( $repository, 'Gtk');


# works shows that GtkActionEntry is deprecated
my N-GObject $base-info = g_irepository_find_by_name( $repository, 'Gtk', 'ActionEntry');
if ?$base-info {
  note 'GtkActionEntry: ', g_base_info_is_deprecated($base-info).Bool;
  g_base_info_unref($base-info)
}

else {
  note 'no base info';
}

# returned base info is a GIObjectInfo
$base-info = g_irepository_find_by_name(
  $repository, 'Gtk', 'Label'
);

note 'GtkLabel: ', g_base_info_is_deprecated($base-info).Bool;
my GEnum $info-type = g_base_info_get_type($base-info);
note 'type: ', $info-type, ' = object';
note 'type name: ', g_object_info_get_type_name($base-info);


my N-GObject $function-info;
my Int $i = g_object_info_get_n_methods($base-info);
for ^$i -> $n {
  $function-info = g_object_info_get_method( $base-info, $n);
  my UInt $flags = g_function_info_get_flags($function-info);
  note "function $n: ",
       g_function_info_get_symbol($function-info),
       ", type $flags.fmt('0b%08b')";

  g_base_info_unref($function-info);
}

g_base_info_unref($base-info);


#===============================================================================
# gi repository
# returns a GIRepository
sub g_irepository_get_default ( --> N-GObject )
  is native(GIREPOSITORY)
  { * }

# returns a GITypelib
sub g_irepository_require (
  N-GObject $repository, gchar-ptr $namespace, gchar-ptr $version,
  guint $flags, CArray[N-GError]
  --> N-GObject
) is native(GIREPOSITORY)
  { * }

# returns a GIBaseInfo
sub g_irepository_find_by_name (
  N-GObject $repository, gchar-ptr $namespace, gchar-ptr $name --> N-GObject
) is native(GIREPOSITORY)
  { * }

sub g_irepository_get_shared_library (
  N-GObject $repository, gchar-ptr $namespace --> gchar-ptr
) is native(GIREPOSITORY)
  { * }

sub g_irepository_get_version (
  N-GObject $repository, gchar-ptr $namespace --> gchar-ptr
) is native(GIREPOSITORY)
  { * }



# gi base info
sub g_base_info_is_deprecated ( N-GObject $info --> gboolean )
  is native(GIREPOSITORY)
  { * }

sub g_base_info_unref ( N-GObject $info )
  is native(GIREPOSITORY)
  { * }

# returns a GIInfoType
sub g_base_info_get_type ( N-GObject $info --> GEnum )
  is native(GIREPOSITORY)
  { * }



# gi object info
sub g_object_info_get_type_name ( N-GObject --> gchar-ptr )
  is native(GIREPOSITORY)
  { * }

# returns a GIFunctionInfo
sub g_object_info_get_method ( N-GObject $info, gint $n --> N-GObject)
  is native(GIREPOSITORY)
  { * }




# function info
sub g_object_info_get_n_methods ( N-GObject $finfo --> gint )
  is native(GIREPOSITORY)
  { * }

sub g_function_info_get_symbol ( N-GObject $finfo --> gchar-ptr )
  is native(GIREPOSITORY)
  { * }

sub g_function_info_get_flags ( N-GObject $finfo --> GEnum )
  is native(GIREPOSITORY)
  { * }
```

**Comment by MARTIMM** [comment]
**There is a compile time penalty on every run of the application, because the modules, classes and types would need to be constructed. But only those elements that are actually used would need to be compiled**

Yes, that is true and that will stay while the pre-generated stuff is compiled once when `use`d. All classes are separate so only those needed are imported into the program.

**On the other hand the multiple hour compile times Xliff repeatedly reports for compilation of his Gtk wrapper is just as unfeasible. From a quick test on how long Gnome::Gtk3 takes to build, it's a lot better.**

That's nice to know. I knew that he was also working on Gtk libraries but didn't know about this problem.

**So, in short, I'm not opposed at all to the pre-compiled binding approach.**

I will continue this approach and maybe later, just out of curiosity I may have some attempt to see what it can bring. maybe a kind of mix between the two.

Thanks anyway for your interest in this.

**Comment by patrickbkr** [comment]
> Yes, that is true and that will stay while the pre-generated stuff is compiled once when `use`d. All classes are separate so only those needed are imported into the program.

But when installing the module, zef will compile all source files. (Without `--/test` they are compiled twice even.) That's where the long compile times happen for Xliff. Once the module is installed, no compilation happens at all when using the pre-generated sources approach.

> I will continue this approach and maybe later, just out of curiosity I may have some attempt to see what it can bring. maybe a kind of mix between the two.

Irrespective of the pre-compiled vs runtime-generated discussion, I favor basing a binding on GIR (either the XML files or libgirepsitory), since once the binding generator works, we basically get access to all the libraries in the Gnome ecosystem and upgrading to newer versions is then also for free. Gtk4 yay!

> Thanks anyway for your interest in this.

Thanks for the time you invested and continuecto invest in your bindings. I really appreciate your work!

**Comment by Leont** [comment]
> Irrespective of the pre-compiled vs runtime-generated discussion, I favor basing a binding on GIR (either the XML files or libgirepsitory), since once the binding generator works, we basically get access to all the libraries in the Gnome ecosystem and upgrading to newer versions is then also for free.

Exactly this. On my computer there are some 271 of them, many of them useful libraries (e.g. GStreamer, WebKit, JavaScriptCore, PolicyKit, DBus)

**Comment by MARTIMM** [comment]
@Patrickbkr, @Leont
> But when installing the module, zef will compile all source files. 
> (Without |--/test| they are compiled twice even.)
Good that you have mentioned this again. Of course I know about it but 
did not thought about it in this discussion. I only thought about the 
situation that the installed packages are used in a new user program. 
So, yes, installing takes a long time and it gets worse when modules and 
packages are added. A good reason to think it all over.

Anyway the techniques used to generate the stuff is not thrown away. The 
documentation is still useful and the way modules are generated can 
maybe turned into an on the fly calling process. But enumerations, 
constants and structures may still have to be generated so the user can 
have symbols instead of using integer numbers.

Well, it is food for the brain, although I must take care not to 
overheat that instrument!

**Comment by patrickbkr** [comment]
>It is already many months ago that we had this discussion so I thought to give some update.

I've looked at the repo from time to time and am impressed by your progress.

>This project is then sitting in between, Using the gir data described in XML files and using that info to build the Raku classes. This will also relieve the user of having the C-libraries installed on their computers.

But they'll still need the XML files which I suspect are usually part of the same package. Also one additionally needs the libxml2 native library. But I don't think this is much of a problem. The approach is still feasible.

>* Is it really faster when I get the info directly from the C-libraries using several calls to get the info needed (now already gathered in the Hash) and then still have to get to the address of the function? I guess not.

I think it doesn't matter much. It just needs to be fast enough. This is a UI library, performance only matters to some extent.

>  * Does the use of FALLBACK have much runtime influence?

I think the new dispatch mechanism now part of MoarVM has improved performance of FALLBACK tremendously.

**Comment by MARTIMM** [comment]
@patrickbkr Hi, thanks for your comment, it is very much appreciated. 

> But they'll still need the XML files which I suspect are usually part of the same package

I see that I was not clear enough on the process. First, a skimmer distills information from the XML files. Then, a  generator creates the raku classes, documentation, and tests from that data. That is happening on my computer only. Some hand work is needed afterward. After that, the finished raku modules can be uploaded and installed by the user, no XML is needed then.

>I think the new dispatch mechanism now part of MoarVM has improved performance of FALLBACK tremendously.

You're right because, in the meantime, I've done tests to write out methods and compared the run-time results of the two implementations (the other making use of FALLBACK). There isn't much difference between the two and writing out all methods would perhaps make a greater impact on compile-time. This means that I stick to the current way of generating the modules.

In that test, I've also seen that the mapping of the functions to the native function address takes much more time and therefore it has been a good choice to postpone that to runtime. All in all, not all functions are always needed, some may never.


test | max | min | mean |
|--|--|--|--|
| use fallback | 0.007366712 | 0.000341382 | 0.000393
| | 0.008176582 | 0.000369133 | 0.000426
| implement methods  | 0.002828965 | 0.000301957 | 0.000344
| | 0.003227472 | 0.000308849 | 0.000351

This table shows a benchmark of two implementations and is run 10.000 times. Before the benchmark is started, both tests were called once to get the native functions ready. So here you won't see that in the table. Those times are some tenths of a second to get three native functions.

**Comment by patrickbkr** [comment]
> @patrickbkr Hi, thanks for your comment, it is very much appreciated.
> 
> > But they'll still need the XML files which I suspect are usually part of the same package
> 
> I see that I was not clear enough on the process. First, a skimmer distills information from the XML files. Then, a generator creates the raku classes, documentation, and tests from that data. That is happening on my computer only. Some hand work is needed afterward. After that, the finished raku modules can be uploaded and installed by the user, no XML is needed then.

Does this mean, that every GIR library will need a corresponding Raku module? And that every new ABI incompatible release of any such library will need a new version of the Raku module?

Do you think it's possible to also have a generic module with which I can bind to any GIR library directly? (That's how the GIR C library is meant to be used. The library provider provides the compiled GIR repository file (and the XML file as well) and a binding module doesn't need to know anything about a specific library, but just work off of the files the library provides.)

**Comment by MARTIMM** [comment]
> Does this mean, that every GIR library will need a corresponding Raku module?

I only have one GIR repository lib on my system. The raku implementation has several distributions with their modules, documentation, and tests. 

> And that every new ABI incompatible release of any such library will need a new version of the Raku module?

Every incompatible change in their ABI will need a change in the raku modules. I assume (guessed from past history) that these changes will not be done quickly and with deprecation phases because the developers will stir up a large code base in C and other language bindings (There is still a Gtk version 2!). 

> Do you think it's possible to also have a generic module with which I can bind to any GIR library directly? 

I am not sure if that is possible because I have some work after the generation of code, documentation, and tests. Code has the least of changes afterward so I could skip the tests and documentation 🤔?
But … if this raku code could follow the GIR library that way, the raku users of this raku distribution may suddenly find that their raku code doesn't work anymore when they are updating the GIR library indirectly by a system update for example. How would they find the problem? I also think that it isn't even possible to warn the user of upcoming changes.

For the moment I think there are too many drawbacks to have it implemented that way. I certainly want to extend my tests on this but not at this moment.

**Comment by patrickbkr** [comment]
Just out of curiosity: What's the current state of things? I skim the commit history from time to time and see loads of commits, so you are hard at work. But I have little idea of the overall plan.

**Comment by MARTIMM** [comment]
Hello Patrick,
I think it is going well enough to say that this project is going in the right direction. There is still some stepping back and forth to decide what to implement and what not. Now I think, not to implement Cairo because of two reasons;
* A [Gnome Blog](https://blog.gtk.org/2023/09/19/paths-in-gtk/) states in its first lines "*It is no secret that we want to get rid of Cairo as the drawing API in GTK, so we can move more of our drawing onto the GPU.*".
* Using the XML parts of GIR, I've seen that only some enumerations are defined but not the API to functions and structures. I would have to use my old programs to read from the C sources directly.

So I am now exploring `Graphene` and `Gsk` to do the things needed. The methods having the Cairo types will be disabled.

I had also to rewrite the generator for structures and unions because the generated code had circular dependencies. And that, in turn, delivered another problem. This is solved but I need another rewrite on the generator. Nothing fancy but it needs to be done.

In the meantime, I made a program based on my Gtk 4 libraries and I am satisfied with the results. I like playing jigsaw puzzles and am used to playing with Palapeli. I didn't like the display of all the puzzles, small pics, no categories, and no progress visible. So I made a puzzle table and after selecting a puzzle the program runs Palapeli to play.

![Screenshot_20240210_144811](https://github.com/MARTIMM/gnome-source-skim-tool/assets/6500673/b64a2bfc-dc05-4d82-911c-02caf094f47c)

Everything is still slow when everything must be compiled. It is quite a beast all those distros together. But after that, it starts much faster. E.g. the puzzle program needs about 7 seconds to start up.
I also must mention that there are a few quirks in Gtk4. not all CSS works and there are a few bugs here and there. Also, I have generated modules of classes that are already deprecated. I need to get that information from elsewhere because it is not (yet) mentioned in the GIR info.

**Comment by patrickbkr** [comment]
Hey! Long time no read! I'm just curious, how are things going? Looking at all the progress since Feburary, when do you think it's the right time for an ecosystem release of Gnome::Gtk4?

**Comment by MARTIMM** [comment]
Hoi Patrick,

Well, I would like to send the stuff in the world. It is something that I don't know if I should write first something about it before releasing. It is a bit different from the previous Gnome::Gtk3 releases. I can at least write in the README with pointers to something bigger.

**Comment by MARTIMM** [comment]
Made libraries available;
```
> fez list
>>= Gnome::Gdk4:ver<0.1.14>:auth<zef:martimm>
>>= Gnome::GdkPixbuf:ver<0.1.2>:auth<zef:martimm>
>>= Gnome::Gio:ver<0.1.22>:auth<zef:martimm>
>>= Gnome::Glib:ver<0.1.12>:auth<zef:martimm>
>>= Gnome::GObject:ver<0.1.11>:auth<zef:martimm>
>>= Gnome::Graphene:ver<0.1.8>:auth<zef:martimm>
>>= Gnome::Gsk4:ver<0.1.5>:auth<zef:martimm>
>>= Gnome::Gtk4:ver<0.1.72>:auth<zef:martimm>
>>= Gnome::N:ver<0.1.38>:auth<zef:martimm>
>>= Gnome::Pango:ver<0.1.9>:auth<zef:martimm>
```
The readme of **Gnome::Gtk4** has an example of how to get started.

**Comment by patrickbkr** [comment]
Wow! I'm really impressed by what you have created. Thank you so much!

**Comment by patrickbkr** [comment]
If you feel up to it, could you write a blog post about this? I guess this needs more visibility.

**Comment by patrickbkr** [comment]
Hey! Just a quick note that the Raku Advent calendar is up.
If you feel that the Gnome libraries are up for it (I believe they are!), this would be a nice chance to promote them a bit more. Maybe showcase one of your programs using the libraries?

**Comment by patrickbkr** [comment]
Oh wow! I missed that entirely, sorry! Then you can pretty much disregard my comment. If I could wish for anything more, I'd love an example of how to use the builder for RAD style development. But don't feel pressured in any way.

**Comment by MARTIMM** [comment]
@patrickbkr 
I was wondering, if you already have some experience with the distros. Is there anything I should have explained a bit more?

**Comment by patrickbkr** [comment]
Sadly I don't. I'd really like to build something with it, but the most pressing projects of mine sadly don't have a GUI need. Once I do get a chance to use it, I'll definitely report back.
